File: app/src/main/java/com/sortedqueue/programmercreek/fragments/NewFillBlankFragment.java
package com.sortedqueue.programmercreek.fragments;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.database.DatabaseError;
import com.sortedqueue.programmercreek.CreekApplication;
import com.sortedqueue.programmercreek.R;
import com.sortedqueue.programmercreek.activity.ProgramListActivity;
import com.sortedqueue.programmercreek.adapter.MatchOptionsDragAdapter;
import com.sortedqueue.programmercreek.adapter.MatchQuestionsDropAdapter;
import com.sortedqueue.programmercreek.constants.ProgrammingBuddyConstants;
import com.sortedqueue.programmercreek.database.CreekUserStats;
import com.sortedqueue.programmercreek.database.ProgramIndex;
import com.sortedqueue.programmercreek.database.ProgramTable;
import com.sortedqueue.programmercreek.database.firebase.FirebaseDatabaseHandler;
import com.sortedqueue.programmercreek.interfaces.ModuleDetailsScrollPageListener;
import com.sortedqueue.programmercreek.interfaces.TestCompletionListener;
import com.sortedqueue.programmercreek.interfaces.WizardNavigationListener;
import com.sortedqueue.programmercreek.util.AnimationUtils;
import com.sortedqueue.programmercreek.util.AuxilaryUtils;
import com.sortedqueue.programmercreek.util.CommonUtils;
import com.startapp.android.publish.adsCommon.Ad;
import com.startapp.android.publish.adsCommon.StartAppAd;
import com.startapp.android.publish.adsCommon.VideoListener;
import com.startapp.android.publish.adsCommon.adListeners.AdEventListener;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Alok on 23/08/17.
 */

public class NewFillBlankFragment extends Fragment implements View.OnClickListener, TestCompletionListener {

    @BindView(R.id.questionRecyclerView)
    RecyclerView questionRecyclerView;
    @BindView(R.id.optionRecyclerView)
    RecyclerView optionRecyclerView;
    List<ProgramTable> mProgramTableList;
    @BindView(R.id.checkButton)
    Button checkButton;
    @BindView(R.id.optionsTextView)
    TextView optionsTextView;
    private Bundle newProgramActivityBundle;
    private int mInvokeMode;
    private ArrayList<ProgramTable> program_TableList;
    private ProgramIndex mProgramIndex;
    private boolean mWizard;
    private ArrayList<ProgramTable> mProgramQuestionList;
    private ArrayList<String> mOptionsList;
    private MatchQuestionsDropAdapter matchQuestionsAdapter;
    private MatchOptionsDragAdapter matchOptionsAdapter;
    private boolean quizComplete;
    private ArrayList<String> solutionList;
    private boolean isAnswered;
    private CreekUserStats creekUserStats;
    private int program_index;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View fragmentView = inflater.inflate(R.layout.fragment_new_fill_blank, container, false);
        ButterKnife.bind(this, fragmentView);
        return fragmentView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        checkButton.setOnClickListener(this);
        mInvokeMode = newProgramActivityBundle.getInt(ProgrammingBuddyConstants.KEY_INVOKE_TEST, -1);
        program_TableList = newProgramActivityBundle.getParcelableArrayList(ProgrammingBuddyConstants.KEY_USER_PROGRAM);
        showHelperDialog();
        if (program_TableList != null && program_TableList.size() > 0) {
            mProgramIndex = (ProgramIndex) newProgramActivityBundle.get(ProgrammingBuddyConstants.KEY_PROG_ID);
            mWizard = newProgramActivityBundle.getBoolean(ProgramListActivity.KEY_WIZARD, false);
            initUI(program_TableList);
        } else {
            if (mInvokeMode == ProgrammingBuddyConstants.KEY_LESSON) {
                mWizard = false;
                new FirebaseDatabaseHandler(getContext()).getProgramIndexInBackGround(newProgramActivityBundle.getInt(ProgrammingBuddyConstants.KEY_PROG_ID),
                        new FirebaseDatabaseHandler.GetProgramIndexListener() {
                            @Override
                            public void onSuccess(ProgramIndex programIndex) {
                                mProgramIndex = programIndex;
                                getProgramTables();
                            }

                            @Override
                            public void onError(DatabaseError databaseError) {
                                CommonUtils.displayToast(getContext(), R.string.unable_to_fetch_data);
                            }
                        });
            } else {
                mProgramIndex = (ProgramIndex) newProgramActivityBundle.get(ProgrammingBuddyConstants.KEY_PROG_ID);
                mWizard = newProgramActivityBundle.getBoolean(ProgramListActivity.KEY_WIZARD, false);

                getProgramTables();
            }
        }


    }

    private void showHelperDialog() {
        AuxilaryUtils.displayInformation(getContext(), R.string.fill_blanks, R.string.match_maker_new_description, new DialogInterface.OnDismissListener() {
            @Override
            public void onDismiss(DialogInterface dialogInterface) {

            }

        });
    }

    private void getProgramTables() {
        new FirebaseDatabaseHandler(getContext()).getProgramTablesInBackground(mProgramIndex.getProgram_index(), new FirebaseDatabaseHandler.GetProgramTablesListener() {
            @Override
            public void onSuccess(ArrayList<ProgramTable> programTables) {
                program_TableList = programTables;
                initUI(program_TableList);
            }

            @Override
            public void onError(DatabaseError databaseError) {
                CommonUtils.displaySnackBar(getActivity(), R.string.unable_to_fetch_data);

            }
        });
    }

    public void setBundle(Bundle bundle) {
        this.newProgramActivityBundle = bundle;
    }

    private void initUI(ArrayList<ProgramTable> program_TableList) {
        questionRecyclerView.setVisibility(View.INVISIBLE);
        optionRecyclerView.setVisibility(View.INVISIBLE);
        optionsTextView.setVisibility(View.INVISIBLE);
        solutionList = new ArrayList<>();
        for (ProgramTable programTable : program_TableList) {
            solutionList.add(programTable.getProgram_Line());
        }

        mProgramTableList = program_TableList;

        if (program_TableList != null && program_TableList.size() > 0) {
            //TODO
            getActivity().setTitle("Fill blanks : " + mProgramIndex.getProgram_Description());
            mProgramQuestionList = ProgramTable.getMatchList(program_TableList, new ProgramTable.MatchOptionsListener() {
                @Override
                public void getOptionsList(ArrayList<String> optionsList) {
                    mOptionsList = optionsList;
                }
            });
            for (ProgramTable programTable : mProgramQuestionList) {
                Log.d("Match", "programTable : " + programTable);
            }
            for (String solution : mOptionsList) {
                Log.d("Match", "solution : " + solution);
            }

        }

        Collections.shuffle(mOptionsList);

        questionRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        if (mOptionsList.size() <= 4) {
            optionRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        } else {
            //optionRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), 2, GridLayoutManager.HORIZONTAL, false));
            optionRecyclerView.setLayoutManager(new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.HORIZONTAL));
        }

        matchQuestionsAdapter = new MatchQuestionsDropAdapter(mProgramQuestionList, true);
        questionRecyclerView.setAdapter(matchQuestionsAdapter);
        matchOptionsAdapter = new MatchOptionsDragAdapter(mOptionsList);
        optionRecyclerView.setAdapter(matchOptionsAdapter);

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                AnimationUtils.slideInToRight(questionRecyclerView);
                AnimationUtils.slideInToLeft(optionRecyclerView);
                AnimationUtils.slideInToLeft(optionsTextView);
            }
        }, 400);

    }

    private WizardNavigationListener wizardNavigationListener;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof WizardNavigationListener) {
            wizardNavigationListener = (WizardNavigationListener) context;
        }
    }

    @Override
    public void onClick(View view) {
        if (checkButton.getText().toString().equalsIgnoreCase("Check")) {
            ArrayList<ProgramTable> programTables = matchQuestionsAdapter.getProgramList();
            int rightAnswers = 0;
            for (int i = 0; i < solutionList.size(); i++) {
                programTables.get(i).isCorrect = solutionList.get(i).trim().equals(programTables.get(i).getProgram_Line().trim());
                if( programTables.get(i).isChoice && programTables.get(i).isCorrect ) rightAnswers++;
                //program_TableList.get(i).setProgram_Line(solutionList.get(i));
            }
            quizComplete = true;
            matchQuestionsAdapter.setChecked(true, programTables);
            isAnswered = true;
            String message = "";
            if( rightAnswers == mOptionsList.size() ) {
                message = "Congratulations, you've got it";
            }
            else if( rightAnswers < (mOptionsList.size() / 2) ) {
                message = "You need improvement, try again";
            }
            else {
                message = "Almost perfect, you are few steps away, retry";
            }
            isAnswered = rightAnswers == mOptionsList.size();
            CommonUtils.displaySnackBar(getActivity(), message + ". You scored : " + rightAnswers + "/" + mOptionsList.size());
            if (mWizard) {
                updateCreekStats();
            }
            if( isAnswered ) {

                if( mWizard ) {
                    checkButton.setText("Finish");
                }
                else {
                    checkButton.setText("Proceed");
                }
            }

        } else {
            if( mWizard )
                getActivity().finish();
            else {
                if( moduleDetailsScrollPageListener != null ) {
                    moduleDetailsScrollPageListener.onScrollForward();
                }
                else {
                    getActivity().finish();
                }


            }

        }

    }

    private void updateCreekStats() {
        creekUserStats = CreekApplication.getInstance().getCreekUserStats();
        switch (mProgramIndex.getProgram_Language().toLowerCase()) {
            case "c":
                creekUserStats.addToUnlockedCProgramIndexList(mProgramIndex.getProgram_index() + 1);
                break;
            case "cpp":
            case "c++":
                creekUserStats.addToUnlockedCppProgramIndexList(mProgramIndex.getProgram_index() + 1);
                break;
            case "java":
                creekUserStats.addToUnlockedJavaProgramIndexList(mProgramIndex.getProgram_index() + 1);
                break;
            case "usp":
                creekUserStats.addToUnlockedUspProgramIndexList(mProgramIndex.getProgram_index() + 1);
                break;
        }
        new FirebaseDatabaseHandler(getContext()).writeCreekUserStats(creekUserStats);
    }

    public ArrayList<String> getmProgramList() {
        return solutionList;
    }

    public void onBackPressed() {
        if (!quizComplete) {
            AuxilaryUtils.showConfirmationDialog(getActivity());
        } else {
            getActivity().finish();
        }

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    @Override
    public int isTestComplete() {
        return isAnswered ? ProgrammingBuddyConstants.KEY_FILL_BLANKS : -1;
    }

    public void setWizardMode(boolean wizardMode) {
        this.mWizard = wizardMode;
    }

    public void setmProgram_Index(int program_index) {
        this.program_index = program_index;
    }

    private ModuleDetailsScrollPageListener moduleDetailsScrollPageListener;
    public void setModulteDetailsScrollPageListener(ModuleDetailsScrollPageListener moduleDetailsScrollPageListener) {
        this.moduleDetailsScrollPageListener = moduleDetailsScrollPageListener;
    }

    public void showRewardedVideoDialog() {
       AuxilaryUtils.displayInformation(getContext(), R.string.hint_video, R.string.reward_video_description,
               new DialogInterface.OnClickListener() {
                   @Override
                   public void onClick(DialogInterface dialogInterface, int i) {
                       showRewardedClick();
                   }
               },

        new DialogInterface.OnDismissListener() {
                   @Override
                   public void onDismiss(DialogInterface dialogInterface) {

                   }
               });
    }


    public void showRewardedClick() {
        final StartAppAd rewardedVideo = new StartAppAd(getContext());

        /**
         * This is very important: set the video listener to be triggered after video
         * has finished playing completely
         */
        rewardedVideo.setVideoListener(new VideoListener() {

            @Override
            public void onVideoCompleted() {
                int maxHints = (mOptionsList.size() / 2);
                if( maxHints > 3 ) {
                    maxHints = 3;
                }
                CommonUtils.displaySnackBar(getActivity(), maxHints + " new hints have been added");

                matchQuestionsAdapter.addHints(maxHints);

            }
        });

        /**
         * Load rewarded by specifying AdMode.REWARDED
         * We are using AdEventListener to trigger ad show
         */
        rewardedVideo.loadAd(StartAppAd.AdMode.REWARDED_VIDEO, new AdEventListener() {

            @Override
            public void onReceiveAd(Ad arg0) {
                rewardedVideo.showAd();
            }

            @Override
            public void onFailedToReceiveAd(Ad arg0) {
                /**
                 * Failed to load rewarded video:
                 * 1. Check that FullScreenActivity is declared in AndroidManifest.xml:
                 * See https://github.com/StartApp-SDK/Documentation/wiki/Android-InApp-Documentation#activities
                 * 2. Is android API level above 16?
                 */
                Log.e("MainActivity", "Failed to load rewarded video with reason: " + arg0.getErrorMessage());
            }
        });
    }
}

File: app/src/main/java/com/sortedqueue/programmercreek/fragments/NewFillBlankFragment.kt
package com.sortedqueue.programmercreek.fragments

import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.os.Handler
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast

import com.google.firebase.database.DatabaseError
import com.google.gson.Gson
import com.sortedqueue.programmercreek.CreekApplication
import com.sortedqueue.programmercreek.R
import com.sortedqueue.programmercreek.activity.ProgramListActivity
import com.sortedqueue.programmercreek.adapter.MatchOptionsDragAdapter
import com.sortedqueue.programmercreek.adapter.MatchQuestionsDropAdapter
import com.sortedqueue.programmercreek.constants.ProgrammingBuddyConstants
import com.sortedqueue.programmercreek.database.CreekUserStats
import com.sortedqueue.programmercreek.database.ProgramIndex
import com.sortedqueue.programmercreek.database.ProgramTable
import com.sortedqueue.programmercreek.database.firebase.FirebaseDatabaseHandler
import com.sortedqueue.programmercreek.interfaces.ModuleDetailsScrollPageListener
import com.sortedqueue.programmercreek.interfaces.TestCompletionListener
import com.sortedqueue.programmercreek.interfaces.WizardNavigationListener
import com.sortedqueue.programmercreek.util.AnimationUtils
import com.sortedqueue.programmercreek.util.AuxilaryUtils
import com.sortedqueue.programmercreek.util.CommonUtils
import com.sortedqueue.programmercreek.util.CreekAnalytics
import com.startapp.android.publish.adsCommon.Ad
import com.startapp.android.publish.adsCommon.StartAppAd
import com.startapp.android.publish.adsCommon.VideoListener
import com.startapp.android.publish.adsCommon.adListeners.AdEventListener
import kotlinx.android.synthetic.main.fragment_new_fill_blank.*

import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList
import java.util.Collections




/**
 * Created by Alok on 23/08/17.
 */

class NewFillBlankFragment : Fragment(), View.OnClickListener, TestCompletionListener {

    private var newProgramActivityBundle: Bundle? = null
    private var mInvokeMode: Int = 0
    private var program_TableList: ArrayList<ProgramTable>? = null
    private var mProgramIndex: ProgramIndex? = null
    private var mWizard: Boolean = false
    private var mProgramQuestionList: ArrayList<ProgramTable>? = null
    private var mOptionsList: ArrayList<String>? = null
    private var matchQuestionsAdapter: MatchQuestionsDropAdapter? = null
    private var matchOptionsAdapter: MatchOptionsDragAdapter? = null
    private var quizComplete: Boolean = false
    private var solutionList: ArrayList<String>? = null
    private var isAnswered: Boolean = false
    private var creekUserStats: CreekUserStats? = null
    private var program_index: Int = 0
    private val TAG = NewFillBlankFragment::class.java.simpleName

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val fragmentView = inflater!!.inflate(R.layout.fragment_new_fill_blank, container, false)
        return fragmentView
    }



    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        checkButton!!.setOnClickListener(this)
        mInvokeMode = newProgramActivityBundle!!.getInt(ProgrammingBuddyConstants.KEY_INVOKE_TEST, -1)
        program_TableList = newProgramActivityBundle!!.getParcelableArrayList<ProgramTable>(ProgrammingBuddyConstants.KEY_USER_PROGRAM)
        showHelperDialog()
        if (program_TableList != null && program_TableList!!.size > 0) {
            mProgramIndex = newProgramActivityBundle!!.get(ProgrammingBuddyConstants.KEY_PROG_ID) as ProgramIndex
            mWizard = newProgramActivityBundle!!.getBoolean(ProgramListActivity.KEY_WIZARD, false)
            initUI(program_TableList)
        } else {
            if (mInvokeMode == ProgrammingBuddyConstants.KEY_LESSON) {
                mWizard = false
                FirebaseDatabaseHandler(context).getProgramIndexInBackGround(newProgramActivityBundle!!.getInt(ProgrammingBuddyConstants.KEY_PROG_ID),
                        object : FirebaseDatabaseHandler.GetProgramIndexListener {
                            override fun onSuccess(programIndex: ProgramIndex) {
                                mProgramIndex = programIndex
                                getProgramTables()
                            }

                            override fun onError(databaseError: DatabaseError) {
                                CommonUtils.displayToast(context, R.string.unable_to_fetch_data)
                            }
                        })
            } else {
                mProgramIndex = newProgramActivityBundle!!.get(ProgrammingBuddyConstants.KEY_PROG_ID) as ProgramIndex
                mWizard = newProgramActivityBundle!!.getBoolean(ProgramListActivity.KEY_WIZARD, false)

                getProgramTables()
            }
        }


    }

    private fun showHelperDialog() {
        AuxilaryUtils.displayInformation(context, R.string.fill_blanks, R.string.match_maker_new_description, DialogInterface.OnDismissListener { })
    }

    private fun getProgramTables() {
        FirebaseDatabaseHandler(context).getProgramTablesInBackground(mProgramIndex!!.program_index, object : FirebaseDatabaseHandler.GetProgramTablesListener {
            override fun onSuccess(programTables: ArrayList<ProgramTable>) {
                program_TableList = programTables
                initUI(program_TableList)
            }

            override fun onError(databaseError: DatabaseError?) {
                CommonUtils.displaySnackBar(activity, R.string.unable_to_fetch_data)

            }
        })
    }

    fun setBundle(bundle: Bundle) {
        this.newProgramActivityBundle = bundle
    }

    private var mProgramTableList: ArrayList<ProgramTable> ?= null

    private fun initUI(program_TableList: ArrayList<ProgramTable>?) {
        if (mProgramIndex != null) {
            try {
                CreekAnalytics.logEvent(TAG, JSONObject(Gson().toJson(mProgramIndex)))
            } catch (e: JSONException) {
                e.printStackTrace()
            }

        }
        questionRecyclerView!!.visibility = View.INVISIBLE
        optionRecyclerView!!.visibility = View.INVISIBLE
        optionsTextView!!.visibility = View.INVISIBLE
        solutionList = ArrayList<String>()
        for (programTable in program_TableList!!) {
            solutionList!!.add(programTable.program_Line)
        }

        mProgramTableList = program_TableList

        if (program_TableList != null && program_TableList.size > 0) {
            //TODO
            activity.title = "Fill blanks : " + mProgramIndex!!.program_Description
            mProgramQuestionList = ProgramTable.getMatchList(program_TableList) { optionsList -> mOptionsList = optionsList }
            for (programTable in mProgramQuestionList!!) {
                Log.d("Match", "programTable : " + programTable)
            }
            for (solution in mOptionsList!!) {
                Log.d("Match", "solution : " + solution)
            }

        }

        Collections.shuffle(mOptionsList!!)

        questionRecyclerView!!.layoutManager = LinearLayoutManager(context)
        if (mOptionsList!!.size <= 4) {
            optionRecyclerView!!.layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.HORIZONTAL)
        } else {
            //optionRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), 2, GridLayoutManager.HORIZONTAL, false));
            optionRecyclerView!!.layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.HORIZONTAL)
        }

        matchQuestionsAdapter = MatchQuestionsDropAdapter(mProgramQuestionList!!, true)
        questionRecyclerView!!.adapter = matchQuestionsAdapter
        matchOptionsAdapter = MatchOptionsDragAdapter(mOptionsList!!)
        optionRecyclerView!!.adapter = matchOptionsAdapter

        Handler().postDelayed({
            AnimationUtils.slideInToRight(questionRecyclerView)
            AnimationUtils.slideInToLeft(optionRecyclerView)
            AnimationUtils.slideInToLeft(optionsTextView)
        }, 400)

    }

    private var wizardNavigationListener: WizardNavigationListener? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is WizardNavigationListener) {
            wizardNavigationListener = context
        }
    }

    override fun onClick(view: View) {
        if (checkButton!!.text.toString().equals("Check", ignoreCase = true)) {
            val programTables = matchQuestionsAdapter!!.programList
            var rightAnswers = 0
            for (i in solutionList!!.indices) {
                programTables!![i].isCorrect = solutionList!![i].trim { it <= ' ' } == programTables[i].program_Line.trim { it <= ' ' }
                if (programTables[i].isChoice && programTables[i].isCorrect) rightAnswers++
                //program_TableList.get(i).setProgram_Line(solutionList.get(i));
            }
            quizComplete = true
            matchQuestionsAdapter!!.setChecked(true, programTables!!)
            isAnswered = true
            var message = ""
            if (rightAnswers == mOptionsList!!.size) {
                message = "Congratulations, you've got it"
            } else if (rightAnswers < mOptionsList!!.size / 2) {
                message = "You need improvement, try again"
            } else {
                message = "Almost perfect, you are few steps away, retry"
            }
            isAnswered = rightAnswers == mOptionsList!!.size
            CommonUtils.displaySnackBar(activity, message + ". You scored : " + rightAnswers + "/" + mOptionsList!!.size)
            if (mWizard) {
                updateCreekStats()
            }
            if (isAnswered) {

                if (mWizard) {
                    checkButton!!.text = "Finish"
                } else {
                    checkButton!!.text = "Proceed"
                }
            }
            if (moduleDetailsScrollPageListener != null)
                moduleDetailsScrollPageListener!!.toggleFABDrawable()

        } else {
            if (mWizard)
                activity.finish()
            else {
                if (moduleDetailsScrollPageListener != null) {
                    moduleDetailsScrollPageListener!!.onScrollForward()
                } else {
                    activity.finish()
                }


            }

        }

    }

    private fun updateCreekStats() {
        creekUserStats = CreekApplication.instance.creekUserStats
        when (mProgramIndex!!.program_Language.toLowerCase()) {
            "c" -> creekUserStats!!.addToUnlockedCProgramIndexList(mProgramIndex!!.program_index + 1)
            "cpp", "c++" -> creekUserStats!!.addToUnlockedCppProgramIndexList(mProgramIndex!!.program_index + 1)
            "java" -> creekUserStats!!.addToUnlockedJavaProgramIndexList(mProgramIndex!!.program_index + 1)
            "usp" -> creekUserStats!!.addToUnlockedUspProgramIndexList(mProgramIndex!!.program_index + 1)
        }
        FirebaseDatabaseHandler(context).writeCreekUserStats(creekUserStats!!)
    }

    fun getmProgramList(): ArrayList<String> {
        return solutionList!!
    }


    fun onBackPressed() {
        if (!quizComplete) {
            AuxilaryUtils.showConfirmationDialog(activity)
        } else {
            activity.finish()
        }

    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    override fun isTestComplete(): Int {
        return if (isAnswered) ProgrammingBuddyConstants.KEY_FILL_BLANKS else -1
    }

    fun setWizardMode(wizardMode: Boolean) {
        this.mWizard = wizardMode
    }

    fun setmProgram_Index(program_index: Int) {
        this.program_index = program_index
    }

    private var moduleDetailsScrollPageListener: ModuleDetailsScrollPageListener? = null
    fun setModulteDetailsScrollPageListener(moduleDetailsScrollPageListener: ModuleDetailsScrollPageListener) {
        this.moduleDetailsScrollPageListener = moduleDetailsScrollPageListener
    }

    fun addHintsToBlanks() {
        /*int maxHints = (mOptionsList.size() / 2);
        if( maxHints > 3 ) {
            maxHints = 3;
        }
        CommonUtils.displaySnackBar(getActivity(), maxHints + " new hints have been added");

        matchQuestionsAdapter.addHints(maxHints);*/
        CommonUtils.displaySnackBar(activity, "Hints have been added")
        matchQuestionsAdapter!!.enableHints()
    }
}
